File: src/main/java/org/willxu/algorithm/service/treenode/InvertBinaryTree.java
/*
 * Copyright (C) 2021  Weiyang(Will) Xu
 *
 * This file is part of algorithm.
 * Algorithm is free software: you can redistribute it and/or modify
 * it under the term of the GNU General Public License version 3 or
 * any later version, as specified in the readme.md file.
 */

package org.willxu.algorithm.service.treenode;

import org.willxu.algorithm.provide.TreeNode;

public interface InvertBinaryTree {
    /**
     * Given the root of a binary tree, invert the tree, and return
     * its root.
     *
     * @param root 0 <= size <= 100
     *             -100 <= Node.val <= 100
     * @return root of invert tree
     */
    TreeNode invertTree(TreeNode root);
}

File: src/leetcode/p0200_0299/p0226_invert_binary_tree/invert_binary_tree.py
from abc import ABC, abstractmethod
from typing import Optional

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class InvertBinaryTree(ABC):
    """
    Given the root of a binary tree, invert the tree, and return its root.
    """
    @abstractmethod
    def invertTree(self, root: Optional[TreeNode]) -> Optional[TreeNode]:
        """
        Args:
            root:
                - The number of nodes in the tree is in the range [0, 100].
                - -100 <= Node.val <= 100
        """
        pass

class InvertBinaryTreeRecursive(InvertBinaryTree):
    def invertTree(self, root: Optional[TreeNode]) -> Optional[TreeNode]:
        if not root:
            return None
        root.left, root.right = root.right, root.left
        self.invertTree(root.left)
        self.invertTree(root.right)
        return root
